python-frontend/src/main/java/org/sonar/python/semantic/v2/types/TrivialTypeInferenceVisitor.java

```diff
@@ -152,6 +156,23 @@ public void visitNumericLiteral(NumericLiteral numericLiteral) {
     numericLiteralImpl.typeV2(new ObjectType(pythonType, new ArrayList<>(), new ArrayList<>()));
   }
 
+  @Override
+  public void visitUnaryExpression(UnaryExpression unaryExpr) {
```

Comment:
I think by having this in the trivial type inference visitor rather than the flow-sensitive one, you are actually not covering cases such as:

```python
x = 42
y = -x
```

The trivial type inference visitor is really meant to evaluate the type of trivial expressions that do not depend on control flow, such as builtins or the names of class and function definitions. It is only used for propagation in some cases where "normal" propagation would not lead to the desired result (see SONARPY-2244 and SONARPY-2245 for an example).

Here, the expression `-42` has a trivial type only because the expression `42` has a trivial type (`INT`), but you don't have a guarantee that the expression for a unary operation will be trivial.